.github/workflows/upstream.yml

```diff
@@ -7,108 +6,137 @@ on:
 
 concurrency:
   group: upstream-${{ github.ref }}
-  cancel-in-progress: true
+  cancel-in-progress: false
 
 env:
-  PIP_ONLY_BINARY: numpy
-  # For cmake:
-  VERBOSE: 1
+  PYTEST_TIMEOUT: 300
 
 jobs:
   standard:
-    name: "🐍 3.11 latest internals • ubuntu-latest • x64"
+    name: "🐍 3.12 latest • ubuntu-latest • x64"
     runs-on: ubuntu-latest
+
+    # Hint: Select the 'python dev' label in the PR web view.
     if: "contains(github.event.pull_request.labels.*.name, 'python dev')"
 
     steps:
+    - name: Show env
+      run: env
+
     - uses: actions/checkout@v3
 
-    - name: Setup Python 3.11
+    - name: Setup Python 3.12
       uses: actions/setup-python@v4
       with:
-        python-version: "3.11-dev"
+        python-version: "3.12-dev"
 
-    - name: Setup Boost (Linux)
-      if: runner.os == 'Linux'
+    - name: Setup Boost
       run: sudo apt-get install libboost-dev
 
     - name: Update CMake
       uses: jwlawson/actions-setup-cmake@v1.13
 
-    - name: Prepare env
+    - name: Run pip installs
       run: |
-        python -m pip install -r tests/requirements.txt
+        python -m pip install --upgrade pip
+        python -m pip install --prefer-binary -r tests/requirements.txt
+    #   python -m pip install --prefer-binary numpy  # SLOW
+    #   python -m pip install --prefer-binary scipy  # FAILED ~Nov 2022
+
+    - name: Show platform info
+      run: python -m platform
 
-    - name: Setup annotations on Linux
-      if: runner.os == 'Linux'
-      run: python -m pip install pytest-github-actions-annotate-failures
+    - name: Show CMake version
+      run: cmake --version
 
-    # First build - C++11 mode and inplace
+    # FIRST BUILD
     - name: Configure C++11
       run: >
-        cmake -S . -B .
+        cmake -S . -B build11
+        -DCMAKE_VERBOSE_MAKEFILE=ON
         -DPYBIND11_WERROR=ON
         -DDOWNLOAD_CATCH=ON
         -DDOWNLOAD_EIGEN=ON
         -DCMAKE_CXX_STANDARD=11
+        -DCMAKE_BUILD_TYPE=Debug
 
     - name: Build C++11
-      run: cmake --build . -j 2
+      run: cmake --build build11 -j 2
 
     - name: Python tests C++11
-      run: cmake --build . --target pytest -j 2
-
-    - name: C++11 tests
-      run: cmake --build .  --target cpptest -j 2
+      run: cmake --build build11 --target pytest -j 2
+
+    # python-3.12.0-alpha.1-linux-22.04-x64.tar.gz  PASSED
+    # python-3.12.0-alpha.2-linux-22.04-x64.tar.gz  FAILED
+    # python-3.12.0-alpha.3-linux-22.04-x64.tar.gz  FAILED
+    #     free(): invalid pointer
+    #     Custom PyConfig
+    #     /home/runner/work/pybind11/pybind11/tests/test_embed/test_interpreter.cpp:175
+    #     /home/runner/work/pybind11/pybind11/tests/test_embed/test_interpreter.cpp:179: FAILED:
+    # gdb traceback points here:
+    #     https://github.com/python/cpython/blame/54289f85b2af1ecf046089ddf535dda1bdf6af24/Python/import.c#L524
+    # - name: C++ tests C++11
+    #   run: cmake --build build11 --target cpptest -j 2
```

Comment:
This is the only 3.12 specific part. I'd rather record logs of things happening outside the repo, certainly not in the CI files. I'd just make an issue and dump this in there. One issue per Python release would make much more sense than renaming a CI file every year. It's easier to update an issue than commit a change to the files.